src/coreclr/vm/jitinterface.cpp

```diff
@@ -1644,49 +1654,82 @@ static uint32_t ThreadLocalOffset(void* p)
     uint8_t* pOurTls = pTls[_tls_index];
     return (uint32_t)((uint8_t*)p - pOurTls);
 }
+#elif TARGET_OSX
+extern "C" void* GetThreadVarsAddress();
 
-void CEEInfo::getThreadLocalStaticBlocksInfo (CORINFO_THREAD_STATIC_BLOCKS_INFO* pInfo, bool isGCType)
+void* getThreadVarsSectionAddressFromDesc(uint8_t* p)
 {
-    CONTRACTL {
-        NOTHROW;
-        GC_NOTRIGGER;
-        MODE_PREEMPTIVE;
-    } CONTRACTL_END;
+    if (!(p[0] == 0x48 && p[1] == 0x8d && p[2] == 0x3d))
+    {
+        // The optimization is disabled if coreclr is not compiled in .so format.
+        return 0;
+    }
 
-    JIT_TO_EE_TRANSITION_LEAF();
+    // At this point, `p` contains the instruction pointer and is pointing to the above opcodes.
+    // These opcodes are patched by the dynamic linker.
+    // Move beyond the opcodes that we have already checked above.
+    p += 3;
 
-    pInfo->tlsIndex.addr = (void*)static_cast<uintptr_t>(_tls_index);
-    pInfo->tlsIndex.accessType = IAT_VALUE;
+    // The descriptor address is located at *p at this point.
+    // (p + 4) below skips the descriptor address bytes embedded in the instruction and
+    // add it to the `instruction pointer` to find out the address.
+    return *(uint32_t*)p + (p + 4);
+}
 
-    pInfo->offsetOfThreadLocalStoragePointer = offsetof(_TEB, ThreadLocalStoragePointer);
-    if (isGCType)
-    {
-        pInfo->offsetOfThreadStaticBlocks = ThreadLocalOffset(&t_GCThreadStaticBlocks);
-        pInfo->offsetOfMaxThreadStaticBlocks = ThreadLocalOffset(&t_GCMaxThreadStaticBlocks);
-    }
-    else
+void* getThreadVarsSectionAddress()
+{
+#ifdef TARGET_ARM64
+    return reinterpret_cast<void*>(GetThreadVarsAddress());
+#else
+    // On x64, the address is related to rip, so, disassemble the function,
+    // read the offset, and then relative to the IP, find the final address of
+    // __thread_vars section.
+    uint8_t* p = reinterpret_cast<uint8_t*>(&GetThreadVarsAddress);
+    return getThreadVarsSectionAddressFromDesc(p);
+#endif // TARGET_ARM64
+}
+
+#else
+
+#ifdef HOST_AMD64
```

Comment:
```suggestion
#ifdef TARGET_AMD64
```
We use TARGET_... in most other places in this code. (It does not really matter whether this code uses TARGET or HOST since they are always the same when compiling this file. It would be nice to be consistent though.)